File: 983_minimum_cost_for_tickets.py
class Solution:
    def mincostTickets(self, days: list[int], costs: list[int]) -> int:
        if not len(days) or len(costs) != 3:
            raise Exception("Invalid input.")

        num_days = len(days)
        min_costs = [0] * len(days)

        for i in range(num_days):
            # First index falling outside of 1-day range.
            j = i - 1

            # First index falling outside of 7-day range.
            k = j
            while k >= 0 and days[i] - days[k] < 7:
                k -= 1

            # First index falling outside of 30-day range.
            l = k
            while l >= 0 and days[i] - days[l] < 30:
                l -= 1

            cost_1 = min_costs[j] + costs[0] if j >= 0 else costs[0]
            cost_2 = min_costs[k] + costs[1] if k >= 0 else costs[1]
            cost_3 = min_costs[l] + costs[2] if l >= 0 else costs[2]

            min_costs[i] = min(cost_1, cost_2, cost_3)

        return min_costs[-1]

File: problems/983-minimum-cost-for-tickets.ts
function mincostTickets(days: number[], costs: number[]): number {
  if (!days.length || costs.length !== 3) {
    throw new Error("Invalid input.")
  }

  const minCosts: number[] = Array(days.length)

  for (let i = 0; i < minCosts.length; i++) {
    // First index falling outside of 1-day range.
    const j = i - 1

    // First index falling outside of 7-day range.
    let k = j
    while (k >= 0 && days[i] - days[k] < 7) {
      k--
    }

    // First index falling outside of 30-day range.
    let l = k
    while (l >= 0 && days[i] - days[l] < 30) {
      l--
    }

    const cost1 = j >= 0 ? minCosts[j] + costs[0] : costs[0]
    const cost2 = k >= 0 ? minCosts[k] + costs[1] : costs[1]
    const cost3 = l >= 0 ? minCosts[l] + costs[2] : costs[2]

    minCosts[i] = Math.min(cost1, cost2, cost3)
  }

  return minCosts.at(-1)!
}
